# Capstone project - Data Exploration

In [1]:
%%html
<style>
    @import url('https://fonts.googleapis.com/css2?family=Raleway:wght@500&display=swap');
    @import url('https://fonts.googleapis.com/css2?family=Roboto:wght@300&display=swap');
    
    .text_cell.rendered
    .rendered_html {
        font-family: 'Roboto';
        font-style: normal;
        font-weight: 300;
        /*color: #1e272e;*/
        letter-spacing: 0.5px;
        font-size: 16px;
        line-height: 1.7em;
    }
    
    .rendered_html h1, h2, h3{
        font-family: 'Raleway';
        font-style: normal;
        font-weight: 500;
        /*color: #ef5762;*/
    }

    div.output_area pre {
        font-family: SFMono-Regular,Consolas,Liberation Mono,Menlo,monospace;
        font-size: 13px;
    }
</style>

In [2]:
import pandas as pd

In [3]:
!ls data

annual-co2-emissions-per-country.csv  population_total.csv
Foreign_Exchange_Rates.csv	      stani4.csv
GlobalLandTemperaturesByCountry.csv


## 1. First contact with the data

In [4]:
emissions = pd.read_csv("data/annual-co2-emissions-per-country.csv")
print(emissions.shape)
emissions.head()

(20853, 4)


Entity Code  Year  Annual CO₂ emissions (tonnes)
0  Afghanistan  AFG  1949                        14656.0
1  Afghanistan  AFG  1950                        84272.0
2  Afghanistan  AFG  1951                        91600.0
3  Afghanistan  AFG  1952                        91600.0
4  Afghanistan  AFG  1953                       106256.0

In [5]:
temperatures = pd.read_csv("data/GlobalLandTemperaturesByCountry.csv")
print(temperatures.shape)
temperatures.head()

(577462, 4)


dt  AverageTemperature  AverageTemperatureUncertainty Country
0  1743-11-01               4.384                          2.294   Åland
1  1743-12-01                 NaN                            NaN   Åland
2  1744-01-01                 NaN                            NaN   Åland
3  1744-02-01                 NaN                            NaN   Åland
4  1744-03-01                 NaN                            NaN   Åland

In [6]:
population_total = pd.read_csv("data/population_total.csv")
print(population_total.shape)
population_total.head()

(195, 302)


country     1800     1801     1802     1803     1804     1805     1806  \
0  Afghanistan  3280000  3280000  3280000  3280000  3280000  3280000  3280000   
1      Albania   400000   402000   404000   405000   407000   409000   411000   
2      Algeria  2500000  2510000  2520000  2530000  2540000  2550000  2560000   
3      Andorra     2650     2650     2650     2650     2650     2650     2650   
4       Angola  1570000  1570000  1570000  1570000  1570000  1570000  1570000   

      1807     1808  ...       2091       2092       2093       2094  \
0  3280000  3280000  ...   76600000   76400000   76300000   76100000   
1   413000   414000  ...    1330000    1300000    1270000    1250000   
2  2560000  2570000  ...   70400000   70500000   70500000   70600000   
3     2650     2650  ...      63000      62900      62900      62800   
4  1570000  1570000  ...  168000000  170000000  172000000  175000000   

        2095       2096       2097       2098       2099       2100  
0   76000000   75800000   75600000   75400000   75200000   74900000  
1    1220000    1190000    1170000    1140000    1110000    1090000  
2   70700000   70700000   70700000   70700000   70700000   70700000  
3      62700      62700      62600      62500      62500      62400  
4  177000000  179000000  182000000  184000000  186000000  188000000  

[5 rows x 302 columns]

In [7]:
currencies = pd.read_csv("data/Foreign_Exchange_Rates.csv")
print(currencies.shape)
currencies.head()

(5217, 24)


Unnamed: 0  Time Serie AUSTRALIA - AUSTRALIAN DOLLAR/US$  \
0           0  2000-01-03                            1.5172   
1           1  2000-01-04                            1.5239   
2           2  2000-01-05                            1.5267   
3           3  2000-01-06                            1.5291   
4           4  2000-01-07                            1.5272   

  EURO AREA - EURO/US$ NEW ZEALAND - NEW ZELAND DOLLAR/US$  \
0               0.9847                              1.9033   
1                 0.97                              1.9238   
2               0.9676                              1.9339   
3               0.9686                              1.9436   
4               0.9714                               1.938   

  UNITED KINGDOM - UNITED KINGDOM POUND/US$ BRAZIL - REAL/US$  \
0                                    0.6146             1.805   
1                                    0.6109            1.8405   
2                                    0.6092             1.856   
3                                     0.607              1.84   
4                                    0.6104             1.831   

  CANADA - CANADIAN DOLLAR/US$ CHINA - YUAN/US$  \
0                       1.4465           8.2798   
1                       1.4518           8.2799   
2                       1.4518           8.2798   
3                       1.4571           8.2797   
4                       1.4505           8.2794   

  HONG KONG - HONG KONG DOLLAR/US$  ... SINGAPORE - SINGAPORE DOLLAR/US$  \
0                           7.7765  ...                           1.6563   
1                           7.7775  ...                           1.6535   
2                            7.778  ...                            1.656   
3                           7.7785  ...                           1.6655   
4                           7.7783  ...                           1.6625   

  DENMARK - DANISH KRONE/US$ JAPAN - YEN/US$ MALAYSIA - RINGGIT/US$  \
0                      7.329           101.7                    3.8   
1                      7.218          103.09                    3.8   
2                      7.208          103.77                    3.8   
3                     7.2125          105.19                    3.8   
4                     7.2285          105.17                    3.8   

  NORWAY - NORWEGIAN KRONE/US$ SWEDEN - KRONA/US$  \
0                        7.964              8.443   
1                        7.934               8.36   
2                        7.935              8.353   
3                         7.94             8.3675   
4                        7.966              8.415   

  SRI LANKA - SRI LANKAN RUPEE/US$ SWITZERLAND - FRANC/US$  \
0                             72.3                  1.5808   
1                            72.65                  1.5565   
2                            72.95                  1.5526   
3                            72.95                   1.554   
4                            73.15                  1.5623   

  TAIWAN - NEW TAIWAN DOLLAR/US$ THAILAND - BAHT/US$  
0                          31.38               36.97  
1                           30.6               37.13  
2                           30.8                37.1  
3                          31.75               37.62  
4                          30.85                37.3  

[5 rows x 24 columns]

In [8]:
for c in currencies.columns: print(c)

Unnamed: 0
Time Serie
AUSTRALIA - AUSTRALIAN DOLLAR/US$
EURO AREA - EURO/US$
NEW ZEALAND - NEW ZELAND DOLLAR/US$
UNITED KINGDOM - UNITED KINGDOM POUND/US$
BRAZIL - REAL/US$
CANADA - CANADIAN DOLLAR/US$
CHINA - YUAN/US$
HONG KONG - HONG KONG DOLLAR/US$
INDIA - INDIAN RUPEE/US$
KOREA - WON/US$
MEXICO - MEXICAN PESO/US$
SOUTH AFRICA - RAND/US$
SINGAPORE - SINGAPORE DOLLAR/US$
DENMARK - DANISH KRONE/US$
JAPAN - YEN/US$
MALAYSIA - RINGGIT/US$
NORWAY - NORWEGIAN KRONE/US$
SWEDEN - KRONA/US$
SRI LANKA - SRI LANKAN RUPEE/US$
SWITZERLAND - FRANC/US$
TAIWAN - NEW TAIWAN DOLLAR/US$
THAILAND - BAHT/US$


In [9]:
convertible_currencies = [c for c in currencies.columns]
convertible_currencies = convertible_currencies[2:]
convertible_currencies.sort()
for c in convertible_currencies: print(c)

AUSTRALIA - AUSTRALIAN DOLLAR/US$
BRAZIL - REAL/US$
CANADA - CANADIAN DOLLAR/US$
CHINA - YUAN/US$
DENMARK - DANISH KRONE/US$
EURO AREA - EURO/US$
HONG KONG - HONG KONG DOLLAR/US$
INDIA - INDIAN RUPEE/US$
JAPAN - YEN/US$
KOREA - WON/US$
MALAYSIA - RINGGIT/US$
MEXICO - MEXICAN PESO/US$
NEW ZEALAND - NEW ZELAND DOLLAR/US$
NORWAY - NORWEGIAN KRONE/US$
SINGAPORE - SINGAPORE DOLLAR/US$
SOUTH AFRICA - RAND/US$
SRI LANKA - SRI LANKAN RUPEE/US$
SWEDEN - KRONA/US$
SWITZERLAND - FRANC/US$
TAIWAN - NEW TAIWAN DOLLAR/US$
THAILAND - BAHT/US$
UNITED KINGDOM - UNITED KINGDOM POUND/US$


In [10]:
stani4 = pd.read_csv("data/stani4.csv")
print(stani4.shape)
stani4.head()

/home/slimbook/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3063: DtypeWarning: Columns (12,13) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(1000000, 17)


LOCATION          Country   VAR                     Variable     IND  \
0      SVK  Slovak Republic  VALU  Value added, current prices  D90T92   
1      SVK  Slovak Republic  VALU  Value added, current prices  D90T92   
2      SVK  Slovak Republic  VALU  Value added, current prices  D90T92   
3      SVK  Slovak Republic  VALU  Value added, current prices  D90T92   
4      SVK  Slovak Republic  VALU  Value added, current prices  D90T92   

                                            Industry  TIME  Time Unit Code  \
0   Creative, arts and entertainment activities; ...  2005  2005       EUR   
1   Creative, arts and entertainment activities; ...  2006  2006       EUR   
2   Creative, arts and entertainment activities; ...  2007  2007       EUR   
3   Creative, arts and entertainment activities; ...  2008  2008       EUR   
4   Creative, arts and entertainment activities; ...  2009  2009       EUR   

   Unit  PowerCode Code PowerCode Reference Period Code Reference Period  \
0  Euro               6  Millions                   NaN              NaN   
1  Euro               6  Millions                   NaN              NaN   
2  Euro               6  Millions                   NaN              NaN   
3  Euro               6  Millions                   NaN              NaN   
4  Euro               6  Millions                   NaN              NaN   

     Value Flag Codes Flags  
0   488.59        NaN   NaN  
1   592.05        NaN   NaN  
2   831.18        NaN   NaN  
3   575.81        NaN   NaN  
4  1059.43        NaN   NaN

In [11]:
stani4.Unit.unique()

array(['Euro', 'Persons', 'Hours', 'Norwegian Krone',
       'Full time equivalent', 'Index', 'Czech Koruna', 'Swedish Krona',
       'Danish Krone', 'Mexican Peso', 'Jobs', 'Pound Sterling',
       'US Dollar', 'Zloty', 'Swiss Franc', 'Yen', 'Iceland Krona',
       'New Israeli Sheqel', 'Forint', 'Chilean Peso', 'Won',
       'Turkish Lira', 'Canadian Dollar', 'Australian Dollar',
       'New Zealand Dollar'], dtype=object)

In [12]:
stani4["Unit Code"].unique()

array(['EUR', 'PER', 'HOUR', 'NOK', 'FTE', 'IDX', 'CZK', 'SEK', 'DKK',
       'MXN', 'JOB', 'GBP', 'USD', 'PLN', 'CHF', 'JPY', 'ISK', 'ILS',
       'HUF', 'CLP', 'KRW', 'TRY', 'CAD', 'AUD', 'NZD'], dtype=object)

In [13]:
stani4.describe()

TIME            Time  PowerCode Code         Value
count  1000000.000000  1000000.000000  1000000.000000  1.000000e+06
mean      2010.395663     2010.395663        4.804557  2.870183e+06
std          3.323266        3.323266        2.089186  6.109616e+07
min       2005.000000     2005.000000        0.000000 -9.491600e+06
25%       2008.000000     2008.000000        3.000000  9.911151e+01
50%       2010.000000     2010.000000        6.000000  9.966895e+02
75%       2013.000000     2013.000000        6.000000  1.780910e+04
max       2017.000000     2017.000000        6.000000  7.262100e+09

## 2. Stani4 Data Exploration

In [14]:
stani4.dtypes

LOCATION                  object
Country                   object
VAR                       object
Variable                  object
IND                       object
Industry                  object
TIME                       int64
Time                       int64
Unit Code                 object
Unit                      object
PowerCode Code             int64
PowerCode                 object
Reference Period Code     object
Reference Period          object
Value                    float64
Flag Codes                object
Flags                     object
dtype: object

In [15]:
stani4.Country.unique()

array(['Slovak Republic', 'Germany', 'Finland', 'Norway',
       'Czech Republic', 'Netherlands', 'France', 'Sweden', 'Slovenia',
       'Austria', 'Italy', 'Belgium', 'Denmark', 'Portugal', 'Greece',
       'Luxembourg', 'Mexico', 'Spain', 'United Kingdom', 'Estonia',
       'Latvia', 'Poland', 'United States', 'Switzerland', 'Japan',
       'Hungary', 'Iceland', 'Israel', 'Chile', 'Korea', 'Canada',
       'Ireland', 'Turkey', 'Australia', 'New Zealand', 'Lithuania'],
      dtype=object)

In [16]:
stani4.LOCATION.unique()

array(['SVK', 'DEU', 'FIN', 'NOR', 'CZE', 'NLD', 'FRA', 'SWE', 'SVN',
       'AUT', 'ITA', 'BEL', 'DNK', 'PRT', 'GRC', 'LUX', 'MEX', 'ESP',
       'GBR', 'EST', 'LVA', 'POL', 'USA', 'CHE', 'JPN', 'HUN', 'ISL',
       'ISR', 'CHL', 'KOR', 'CAN', 'IRL', 'TUR', 'AUS', 'NZL', 'LTU'],
      dtype=object)

In [17]:
stani4.IND.unique()

array(['D90T92', 'D45', 'D28', 'D84', 'D01', 'D10T12', 'D41T43', 'D03',
       'D36', 'D23', 'D71', 'D19', 'D80T82', 'D95', 'D49', 'D24', 'D29',
       'D16', 'D85', 'D02', 'D59T60', 'D74T75', 'D72', 'D69T70', 'D62T63',
       'D22', 'D53', 'D79', 'D99', 'D94', 'D31T32', 'D55T56', 'D52',
       'D13T15', 'D58', 'D27', 'D18', 'D05T09', 'D73', 'D66', 'D78',
       'D97T98', 'D77', 'D86', 'D51', 'D20', 'D17', 'D93', 'D46', 'D65',
       'D87T88', 'D61', 'D26', 'D25', 'D96', 'D33', 'D64', 'D50',
       'D37T39', 'D68', 'D47', 'D30', 'D21', 'D35', 'D01T03', 'D01T02',
       'D05T39', 'D05T06', 'D07T08', 'D09', 'D10T33', 'D10T11', 'D10',
       'D11', 'D12', 'D13T14', 'D13', 'D14', 'D15', 'D16T18', 'D19T23',
       'D20T21', 'D22T23', 'D24T25', 'D241T31', 'D242T32', 'D26T28',
       'D26T27', 'D29T30', 'D301', 'D303', 'D304', 'D302A9', 'D31T33',
       'D35T39', 'D36T39', 'D45T56', 'D45T47', 'D49T53', 'D58T63',
       'D58T60', 'D64T66', 'D68T82', 'D69T82', 'D69T75', 'D69T71',
       'D73T75

In [18]:
stani4.Industry.unique()

array([' Creative, arts and entertainment activities; libraries, archives, museums and other cultural activities; gambling and betting activities',
       ' Wholesale and retail trade and repair of motor vehicles and motorcycles',
       ' Machinery and equipment n.e.c.',
       ' Public administration and defence; compulsory social security [O]',
       ' Crop and animal production, hunting and related service activities',
       ' Food products, beverages and tobacco', ' Construction [F]',
       ' Fishing and aquaculture',
       ' Water collection, treatment and supply',
       ' Other non-metallic mineral products',
       ' Architectural and engineering activities; technical testing and analysis',
       ' Coke and refined petroleum products',
       ' Security and investigation activities; services to buildings and landscape activities; office administrative, office support and other business support activities',
       ' Repair of computers and personal and household goods',
  

In [19]:
stani4.loc[stani4['Value'].idxmax()]

LOCATION                                                            KOR
Country                                                           Korea
VAR                                                                CAPG
Variable                 Gross capital stock, current replacement costs
IND                                                              D01T99
Industry                                                          Total
TIME                                                               2014
Time                                                               2014
Unit Code                                                           KRW
Unit                                                                Won
PowerCode Code                                                        6
PowerCode                                                      Millions
Reference Period Code                                               NaN
Reference Period                                                

## Important insights

- Data goes from 2005 to 2017.
- Countries included are: Slovak Republic, Germany, Finland, Norway, Czech Republic, Netherlands, France, Sweden, Slovenia, Austria, Italy, Belgium, Denmark, Portugal, Greece, Luxembourg, Mexico, Spain, United Kingdom, Estonia, Latvia, Poland, United States, Switzerland, Japan, Hungary, Iceland, Israel, Chile, Korea, Canada, Ireland, Turkey, Australia, New Zealand and Lithuania.
- There are all types of industries.
- Columns 12 and 13 have mixed data types.
- It's not clear what industries with negative value mean.
- Not all rows have the same currency -> Include a currency conversion dimension.
- Columns: LOCATION, Country, VAR, Variable, IND, Industry, TIME, Time, Unit Code, Unit, PowerCode Code, PowerCode, Reference Period Code, Reference Period, Value, Flag Codes, Flags.

## 3. Country Codes Exploration

In [20]:
stani4.LOCATION.unique()

array(['SVK', 'DEU', 'FIN', 'NOR', 'CZE', 'NLD', 'FRA', 'SWE', 'SVN',
       'AUT', 'ITA', 'BEL', 'DNK', 'PRT', 'GRC', 'LUX', 'MEX', 'ESP',
       'GBR', 'EST', 'LVA', 'POL', 'USA', 'CHE', 'JPN', 'HUN', 'ISL',
       'ISR', 'CHL', 'KOR', 'CAN', 'IRL', 'TUR', 'AUS', 'NZL', 'LTU'],
      dtype=object)

In [21]:
emissions.Code.unique()

array(['AFG', nan, 'ALB', 'DZA', 'AND', 'AGO', 'AIA', 'ATG', 'ARG', 'ARM',
       'ABW', 'AUS', 'AUT', 'AZE', 'BHS', 'BHR', 'BGD', 'BRB', 'BLR',
       'BEL', 'BLZ', 'BEN', 'BMU', 'BTN', 'BOL', 'BES', 'BIH', 'BWA',
       'BRA', 'VGB', 'BRN', 'BGR', 'BFA', 'BDI', 'KHM', 'CMR', 'CAN',
       'CPV', 'CYM', 'CAF', 'TCD', 'CHL', 'CHN', 'CXR', 'COL', 'COM',
       'COG', 'COK', 'CRI', 'CIV', 'HRV', 'CUB', 'CUW', 'CYP', 'CZE',
       'OWID_CZS', 'COD', 'DNK', 'DJI', 'DMA', 'DOM', 'ECU', 'EGY', 'SLV',
       'GNQ', 'ERI', 'EST', 'ETH', 'FRO', 'FLK', 'FJI', 'FIN', 'FRA',
       'GUF', 'PYF', 'GAB', 'GMB', 'GEO', 'DEU', 'GHA', 'GIB', 'GRC',
       'GRL', 'GRD', 'GLP', 'GTM', 'GIN', 'GNB', 'GUY', 'HTI', 'HND',
       'HKG', 'HUN', 'ISL', 'IND', 'IDN', 'IRN', 'IRQ', 'IRL', 'ISR',
       'ITA', 'JAM', 'JPN', 'JOR', 'KAZ', 'KEN', 'KIR', 'KWT', 'KGZ',
       'LAO', 'LVA', 'LBN', 'LSO', 'LBR', 'LBY', 'LIE', 'LTU', 'LUX',
       'MAC', 'MKD', 'MDG', 'MWI', 'MYS', 'MDV', 'MLI', 'MLT', 'MHL',
       'MT

In [22]:
stani4_cc = stani4.LOCATION.unique()
emissions_cc = emissions.Code.unique()
    
for cc in stani4_cc:
    st_c = stani4[stani4.LOCATION == cc].Country.unique()[0]
    em_c = emissions[emissions.Code == cc].Entity.unique()[0]
    print(f'Code: {cc}\t Stani4 Country: {st_c}\t\t Emissions country: {em_c}')

Code: SVK	 Stani4 Country: Slovak Republic		 Emissions country: Slovakia
Code: DEU	 Stani4 Country: Germany		 Emissions country: Germany
Code: FIN	 Stani4 Country: Finland		 Emissions country: Finland
Code: NOR	 Stani4 Country: Norway		 Emissions country: Norway
Code: CZE	 Stani4 Country: Czech Republic		 Emissions country: Czech Republic
Code: NLD	 Stani4 Country: Netherlands		 Emissions country: Netherlands
Code: FRA	 Stani4 Country: France		 Emissions country: France
Code: SWE	 Stani4 Country: Sweden		 Emissions country: Sweden
Code: SVN	 Stani4 Country: Slovenia		 Emissions country: Slovenia
Code: AUT	 Stani4 Country: Austria		 Emissions country: Austria
Code: ITA	 Stani4 Country: Italy		 Emissions country: Italy
Code: BEL	 Stani4 Country: Belgium		 Emissions country: Belgium
Code: DNK	 Stani4 Country: Denmark		 Emissions country: Denmark
Code: PRT	 Stani4 Country: Portugal		 Emissions country: Portugal
Code: GRC	 Stani4 Country: Greece		 Emissions country: Greece
Code: LUX	 Stani4

In [23]:
emissions[emissions.Code == 'OWID_CZS']

Entity      Code  Year  Annual CO₂ emissions (tonnes)
5213  Czechoslovakia  OWID_CZS  1860                       267472.0
5214  Czechoslovakia  OWID_CZS  1861                       879360.0
5215  Czechoslovakia  OWID_CZS  1862                      1494912.0
5216  Czechoslovakia  OWID_CZS  1863                      2110464.0
5217  Czechoslovakia  OWID_CZS  1864                      2726016.0
...              ...       ...   ...                            ...
5340  Czechoslovakia  OWID_CZS  1987                    244440096.0
5341  Czechoslovakia  OWID_CZS  1988                    240559920.0
5342  Czechoslovakia  OWID_CZS  1989                    232667664.0
5343  Czechoslovakia  OWID_CZS  1990                    217579312.0
5344  Czechoslovakia  OWID_CZS  1991                    200655296.0

[132 rows x 4 columns]

In [24]:
emissions[emissions.Code == 'OWID_WRL']

Entity      Code  Year  Annual CO₂ emissions (tonnes)
20353  World  OWID_WRL  1751                   9.350528e+06
20354  World  OWID_WRL  1752                   9.354192e+06
20355  World  OWID_WRL  1753                   9.354192e+06
20356  World  OWID_WRL  1754                   9.357856e+06
20357  World  OWID_WRL  1755                   9.361520e+06
...      ...       ...   ...                            ...
20615  World  OWID_WRL  2013                   3.520789e+10
20616  World  OWID_WRL  2014                   3.550583e+10
20617  World  OWID_WRL  2015                   3.546275e+10
20618  World  OWID_WRL  2016                   3.567510e+10
20619  World  OWID_WRL  2017                   3.615326e+10

[267 rows x 4 columns]

In [25]:
emissions[emissions.Code.isnull()].Entity.unique()

array(['Africa', 'Americas (other)', 'Antarctic Fisheries',
       'Asia and Pacific (other)', 'EU-28', 'Europe (other)',
       'International transport', 'Kyrgysztan', 'Middle East',
       'Statistical differences', 'Wallis and Futuna Islands'],
      dtype=object)

Only rows from emissions with country codes matching industry country codes will be included. 

## 4. Countries exploration

In [26]:
stani4_c = stani4.Country.unique()
temperatures_c = temperatures.Country.unique()
    
for c in stani4_c:
    if c in temperatures_c:
        print(f'Country {c} found in both tables.')
    else:
        print(f'!!Country {c} not found in temperatures table.')

!!Country Slovak Republic not found in temperatures table.
Country Germany found in both tables.
Country Finland found in both tables.
Country Norway found in both tables.
Country Czech Republic found in both tables.
Country Netherlands found in both tables.
Country France found in both tables.
Country Sweden found in both tables.
Country Slovenia found in both tables.
Country Austria found in both tables.
Country Italy found in both tables.
Country Belgium found in both tables.
Country Denmark found in both tables.
Country Portugal found in both tables.
Country Greece found in both tables.
Country Luxembourg found in both tables.
Country Mexico found in both tables.
Country Spain found in both tables.
Country United Kingdom found in both tables.
Country Estonia found in both tables.
Country Latvia found in both tables.
Country Poland found in both tables.
Country United States found in both tables.
Country Switzerland found in both tables.
Country Japan found in both tables.
Country H

In [27]:
import re
for c in temperatures_c:
    if re.compile('South').match(c): print(c)
    if re.compile('Slovak').match(c): print(c)
    if re.compile('North').match(c): print(c)        

North America
North Korea
Northern Mariana Islands
Slovakia
South Africa
South America
South Georgia And The South Sandwich Isla
South Korea


'Slovak Republic' matches 'Slovakia', and we'll asume 'Korea' matches 'South Korea', as it is very unlike that the providers from the industry data had information about North Korean industries.

## 5. Check emissions and temperatures join

In [60]:
print(len(temperatures.Country.unique()))
temperatures.Country.unique()

243


array(['Åland', 'Afghanistan', 'Africa', 'Albania', 'Algeria',
       'American Samoa', 'Andorra', 'Angola', 'Anguilla', 'Antarctica',
       'Antigua And Barbuda', 'Argentina', 'Armenia', 'Aruba', 'Asia',
       'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Baker Island', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium',
       'Belize', 'Benin', 'Bhutan', 'Bolivia',
       'Bonaire, Saint Eustatius And Saba', 'Bosnia And Herzegovina',
       'Botswana', 'Brazil', 'British Virgin Islands', 'Bulgaria',
       'Burkina Faso', 'Burma', 'Burundi', "Côte D'Ivoire", 'Cambodia',
       'Cameroon', 'Canada', 'Cape Verde', 'Cayman Islands',
       'Central African Republic', 'Chad', 'Chile', 'China',
       'Christmas Island', 'Colombia', 'Comoros',
       'Congo (Democratic Republic Of The)', 'Congo', 'Costa Rica',
       'Croatia', 'Cuba', 'Curaçao', 'Cyprus', 'Czech Republic',
       'Denmark (Europe)', 'Denmark', 'Djibouti', 'Dominica',
       'Dominican Republic', 'Ecu

In [59]:
print(len(emissions.Entity.unique()))
emissions.Entity.unique()

233


array(['Afghanistan', 'Africa', 'Albania', 'Algeria', 'Americas (other)',
       'Andorra', 'Angola', 'Anguilla', 'Antarctic Fisheries',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba',
       'Asia and Pacific (other)', 'Australia', 'Austria', 'Azerbaijan',
       'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus',
       'Belgium', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bolivia',
       'Bonaire Sint Eustatius and Saba', 'Bosnia and Herzegovina',
       'Botswana', 'Brazil', 'British Virgin Islands', 'Brunei',
       'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Canada', 'Cape Verde', 'Cayman Islands',
       'Central African Republic', 'Chad', 'Chile', 'China',
       'Christmas Island', 'Colombia', 'Comoros', 'Congo', 'Cook Islands',
       'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Curacao',
       'Cyprus', 'Czech Republic', 'Czechoslovakia',
       'Democratic Republic of Congo', 'Denmark', 'Djibouti', 'Dominica',
       'D

At first glance, it should work well. Let's try to make the join like in sql.py.

In [53]:
import math
countries = emissions[["Entity", "Code"]].copy()
countries = countries.drop_duplicates()
countries = countries.dropna()
countries.columns = ["Country", "Code"]
countries = countries.set_index("Country")
countries

Code
Country              
Afghanistan       AFG
Albania           ALB
Algeria           DZA
Andorra           AND
Angola            AGO
...               ...
Vietnam           VNM
World        OWID_WRL
Yemen             YEM
Zambia            ZMB
Zimbabwe          ZWE

[222 rows x 1 columns]

In [56]:
join = temperatures.join(countries, on="Country", how="left")
join

dt  AverageTemperature  AverageTemperatureUncertainty  \
0       1743-11-01               4.384                          2.294   
1       1743-12-01                 NaN                            NaN   
2       1744-01-01                 NaN                            NaN   
3       1744-02-01                 NaN                            NaN   
4       1744-03-01                 NaN                            NaN   
...            ...                 ...                            ...   
577457  2013-05-01              19.059                          1.022   
577458  2013-06-01              17.613                          0.473   
577459  2013-07-01              17.000                          0.453   
577460  2013-08-01              19.759                          0.717   
577461  2013-09-01                 NaN                            NaN   

         Country Code  
0          Åland  NaN  
1          Åland  NaN  
2          Åland  NaN  
3          Åland  NaN  
4          Åland  NaN  
...          ...  ...  
577457  Zimbabwe  ZWE  
577458  Zimbabwe  ZWE  
577459  Zimbabwe  ZWE  
577460  Zimbabwe  ZWE  
577461  Zimbabwe  ZWE  

[577462 rows x 5 columns]

In [62]:
join[["Country", "Code"]].drop_duplicates().dropna().T

3239     7310     10549    15031   18270     20148      25466   \
Country  Afghanistan  Albania  Algeria  Andorra  Angola  Anguilla  Argentina   
Code             AFG      ALB      DZA      AND     AGO       AIA        ARG   

          27367  30169      34816   ...                546353          551641  \
Country  Armenia  Aruba  Australia  ...  United Arab Emirates  United Kingdom   
Code         ARM    ABW        AUS  ...                   ARE             GBR   

                554880   557821      560002     562339   564616 571879  \
Country  United States  Uruguay  Uzbekistan  Venezuela  Vietnam  Yemen   
Code               USA      URY         UZB        VEN      VNM    YEM   

         573532    575497  
Country  Zambia  Zimbabwe  
Code        ZMB       ZWE  

[2 rows x 188 columns]

We have 188 countries, it worked really well! We should be able to join the tables on redshift without problems. :-)